src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/PartitionedRateLimiter.cs

```diff
@@ -0,0 +1,173 @@
+        }
+
+        private RateLimiter GetRateLimiter(TResource resourceID)
+        {
+            RateLimitPartition<TKey> partition = _partitioner(resourceID);
+            Lazy<RateLimiter>? limiter;
+            lock (Lock)
+            {
+                ThrowIfDisposed();
+                if (!_limiters.TryGetValue(partition.PartitionKey, out limiter))
+                {
+                    // Using Lazy avoids calling user code (partition.Factory) inside the lock
+                    limiter = new Lazy<RateLimiter>(() => partition.Factory(partition.PartitionKey));
+                    _limiters.Add(partition.PartitionKey, limiter);
+                }
+            }
+            return limiter.Value;
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (!disposing)
+            {
+                return;
+            }
+
+            lock (Lock)
+            {
+                if (_disposed)
+                {
+                    return;
+                }
+                _disposed = true;
+
+                _timer?.Dispose();
+
+                foreach (KeyValuePair<TKey, Lazy<RateLimiter>> limiter in _limiters)
+                {
+                    limiter.Value.Value.Dispose();
+                }
+                _limiters.Clear();
+            }
+        }
+
+        protected override async ValueTask DisposeAsyncCore()
+        {
+            Dictionary<TKey, Lazy<RateLimiter>> limiters;
+            lock (Lock)
+            {
+                if (_disposed)
+                {
+                    return;
+                }
+                _disposed = true;
+
+                _timer?.Dispose();
+
+                limiters = _limiters;
+                _limiters = new();
+            }
+
+            foreach (KeyValuePair<TKey, Lazy<RateLimiter>> limiter in limiters)
+            {
+                await limiter.Value.Value.DisposeAsync().ConfigureAwait(false);
+            }
+
+            limiters.Clear();
+
+            return;
+        }
+
+        private void ThrowIfDisposed()
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(PartitionedRateLimiter));
+            }
+        }
+
+        private static void Replenish(object? state)
+        {
+            DefaultPartitionedRateLimiter<TResource, TKey> limiter = (state as DefaultPartitionedRateLimiter<TResource, TKey>)!;
+            Debug.Assert(limiter is not null);
+
+            lock (limiter.Lock)
+            {
+                if (limiter._disposed)
+                {
+                    return;
+                }
+
+                foreach (KeyValuePair<TKey, Lazy<RateLimiter>> kvp in limiter._limiters)
+                {
+                    // Check IsValueCreated to avoid potentially blocking inside the lock if the Lazy hasn't been initialized yet
+                    if (kvp.Value.IsValueCreated)
```

Comment:
This avoids calling the `partition.Factory` inside of a lock, but `ReplenishingRateLimiter.TryReplenish()` `TryReplenish` is also abstract. This reminds me of the deadlock I introduced in ConfigurationManager (#61747). That was caused in part by calling `IConfigurationProvider.Load()` with a lock.

Of course, it took a bunch of other complicated interactions to cause the actual deadlock, but it's best to call user-definable abstract code without a lock. I know we want to avoid copying the limiters here. I think we could amortize this by copying the limiters into a cached `List<RateLimiter>` that's stored in a field.

Reply:
Added a cache and moved `TryReplenish` outside of the lock.